// @snyk/java-call-graph-builder/dist/metrics.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const _ = require("@snyk/lodash");
const metricsState = {
    getEntrypoints: { seconds: 0, nanoseconds: 0 },
    generateCallGraph: { seconds: 0, nanoseconds: 0 },
    mapClassesPerJar: { seconds: 0, nanoseconds: 0 },
    getCallGraph: { seconds: 0, nanoseconds: 0 },
};
function start(metric) {
    const [secs, nsecs] = process.hrtime();
    metricsState[metric] = { seconds: secs, nanoseconds: nsecs };
}
function stop(metric) {
    const { seconds, nanoseconds } = metricsState[metric] || {
        seconds: 0,
        nanoseconds: 0,
    };
    const [secs, nsecs] = process.hrtime([seconds, nanoseconds]);
    metricsState[metric] = { seconds: secs, nanoseconds: nsecs };
}
function getMetrics() {
    return _(metricsState)
        .mapValues(({ seconds, nanoseconds }) => seconds + nanoseconds / 1e9)
        .value();
}
exports.getMetrics = getMetrics;
function timeIt(metric, fn) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        start(metric);
        const x = yield fn();
        stop(metric);
        return x;
    });
}
exports.timeIt = timeIt;
