// @snyk/java-call-graph-builder/dist/mvn-wrapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
require("source-map-support/register");
const sub_process_1 = require("./sub-process");
function getMvnCommandArgsForMvnExec(targetPath) {
    return [
        '-q',
        'exec:exec',
        '-Dexec.classpathScope="compile"',
        '-Dexec.executable="echo"',
        '-Dexec.args="%classpath"',
        '-f',
        targetPath,
    ];
}
function getMvnCommandArgsForDependencyPlugin(targetPath) {
    return ['dependency:build-classpath', '-f', targetPath];
}
function runMvnCommand(mvnCommandArgs, targetPath) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        return sub_process_1.execute('mvn', mvnCommandArgs, { cwd: targetPath });
    });
}
function parseMvnDependencyPluginCommandOutput(mvnCommandOutput) {
    const outputLines = mvnCommandOutput.split('\n');
    const mvnClassPaths = [];
    let startIndex = 0;
    let i = outputLines.indexOf('[INFO] Dependencies classpath:', startIndex);
    while (i > -1) {
        if (outputLines[i + 1] !== '') {
            mvnClassPaths.push(outputLines[i + 1]);
        }
        startIndex = i + 2;
        i = outputLines.indexOf('[INFO] Dependencies classpath:', startIndex);
    }
    return mvnClassPaths;
}
exports.parseMvnDependencyPluginCommandOutput = parseMvnDependencyPluginCommandOutput;
function parseMvnExecCommandOutput(mvnCommandOutput) {
    return mvnCommandOutput.trim().split('\n');
}
exports.parseMvnExecCommandOutput = parseMvnExecCommandOutput;
function mergeMvnClassPaths(classPaths) {
    // this magic joins all items in array with :, splits result by : again
    // makes Set (to uniq items), create Array from it and join it by : to have
    // proper path like format
    return Array.from(new Set(classPaths.join(':').split(':'))).join(':');
}
exports.mergeMvnClassPaths = mergeMvnClassPaths;
function getClassPathFromMvn(targetPath) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        let mvnCommandArgs = [];
        let mvnOutput;
        let classPaths = [];
        try {
            try {
                // there are two ways of getting classpath - either from maven plugin or by exec command
                // try `mvn exec` for classpath
                mvnCommandArgs = getMvnCommandArgsForMvnExec(targetPath);
                mvnOutput = yield runMvnCommand(mvnCommandArgs, targetPath);
                classPaths = parseMvnExecCommandOutput(mvnOutput);
            }
            catch (e) {
                // if it fails, try mvn dependency:build-classpath
                // TODO send error message for further analysis
                mvnCommandArgs = getMvnCommandArgsForDependencyPlugin(targetPath);
                mvnOutput = yield runMvnCommand(mvnCommandArgs, targetPath);
                classPaths = parseMvnDependencyPluginCommandOutput(mvnOutput);
            }
            return mergeMvnClassPaths(classPaths);
        }
        catch (e) {
            throw new Error(`mvn command 'mvn ${mvnCommandArgs.join(' ')} failed with error: ${e}`);
        }
    });
}
exports.getClassPathFromMvn = getClassPathFromMvn;
