// snyk-docker-plugin/dist/ls-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const event_loop_spinner_1 = require("event-loop-spinner");
/**
 * Iterate over all files of a given DiscoveredDirectory structure.
 */
function iterateFiles(root, iterator) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        for (const f of root.files) {
            iterator(f);
        }
        if (event_loop_spinner_1.eventLoopSpinner.isStarving()) {
            yield event_loop_spinner_1.eventLoopSpinner.spin();
        }
        for (const d of root.subDirs) {
            yield iterateFiles(d, iterator);
        }
    });
}
exports.iterateFiles = iterateFiles;
/**
 * Parse the output of a ls command and return a DiscoveredDirectory
 * structure. Can handle plain and recursive output. Root path will
 * always be normalized to '/'. State scope is a directory i.e. the
 * parser can handle missing parent directories or an out-of-order
 * directory sequence.
 */
function parseLsOutput(output) {
    const outputLines = output.trim().split("\n");
    const isRecursiveOutput = outputLines[0].endsWith(":");
    const res = {
        name: "/",
        subDirs: [],
        files: [],
    };
    let pathPrefix = "";
    let currPath = "/";
    let currDir = res;
    outputLines.forEach((i, n) => {
        if (i.endsWith(":")) {
            if (n === 0 && i !== "/:") {
                pathPrefix = i.substring(0, i.length - 1);
            }
            currPath = i.substring(pathPrefix.length, i.length - 1) || "/";
            if (currPath !== "/") {
                currDir = getSubDir(res, currPath.substring(1).split("/"));
            }
            else {
                currDir = res;
            }
        }
        else if (i !== "" && !i.endsWith("/") && i !== "." && i !== "..") {
            currDir.files.push({
                name: i,
                path: currPath,
            });
        }
        else if (!isRecursiveOutput &&
            i.endsWith("/") &&
            i !== "./" &&
            i !== "../") {
            currDir.subDirs.push({
                name: i.substring(0, i.length - 1),
                subDirs: [],
                files: [],
            });
        }
    });
    return res;
}
exports.parseLsOutput = parseLsOutput;
/**
 * Return a subdirectory of a given DiscoveredDirectory. Create
 * parent directories of the subdirectory if necessary.
 */
function getSubDir(root, path) {
    let subdir;
    if (!path || !path.length) {
        return root;
    }
    subdir = root.subDirs.find((i) => i.name === path[0]);
    if (!subdir) {
        subdir = { name: path[0], subDirs: [], files: [] };
        root.subDirs.push(subdir);
    }
    return getSubDir(subdir, path.slice(1));
}
